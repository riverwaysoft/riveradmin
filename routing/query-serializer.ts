import { parse, stringify } from 'qs';

type Window = {
  location: Location;
};

export class QuerySerializer {
  parseQueryParams(_window: Window = window) {
    return parse(_window.location.search, { ignoreQueryPrefix: true });
  }

  stringifyParams(params: object) {
    return stringify(params);
  }
}
