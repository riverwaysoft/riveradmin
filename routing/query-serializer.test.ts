import { QuerySerializer } from './query-serializer';

describe('QuerySerializer', () => {
  const querySerializer = new QuerySerializer();

  it('should parse url from window object', () => {
    const search = '?id=1&date[from]=2015&date[to]=2019&status[0]=a&status[1]=b';
    const windowMock = { location: { search } } as any;

    const filters = querySerializer.parseQueryParams(windowMock);
    expect(filters).toMatchSnapshot();
  });

  it('should support url encoded values', () => {
    const search = '?id=1&date%5Bfrom%5D=2015&date%5Bto%5D=2019&status%5B0%5D=a&status%5B1%5D=b';
    const windowMock = { location: { search } } as any;

    const filters = querySerializer.parseQueryParams(windowMock);
    expect(filters).toMatchSnapshot();
  });

  it('should parse empty query param list', () => {
    const windowMock = { location: { search: '' } } as any;

    const filters = querySerializer.parseQueryParams(windowMock);
    expect(filters).toMatchSnapshot();
  });

  it('should parse query prefix only', () => {
    const windowMock = { location: { search: '?' } } as any;

    const filters = querySerializer.parseQueryParams(windowMock);
    expect(filters).toMatchSnapshot();
  });

  it('should stringify complex filter', () => {
    const filter = {
      test: 1,
      range: {
        min: 10,
        max: 20,
      },
      'partner.companyName': 'someName',
    };

    const serialized = querySerializer.stringifyParams(filter);
    expect(serialized).toMatchSnapshot();

    const windowMock = { location: { search: serialized } } as any;
    expect(querySerializer.parseQueryParams(windowMock)).toMatchSnapshot();
  });
});
