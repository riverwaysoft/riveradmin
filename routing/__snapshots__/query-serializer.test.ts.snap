// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuerySerializer should parse empty query param list 1`] = `Object {}`;

exports[`QuerySerializer should parse query prefix only 1`] = `Object {}`;

exports[`QuerySerializer should parse url from window object 1`] = `
Object {
  "date": Object {
    "from": "2015",
    "to": "2019",
  },
  "id": "1",
  "status": Array [
    "a",
    "b",
  ],
}
`;

exports[`QuerySerializer should stringify complex filter 1`] = `"test=1&range%5Bmin%5D=10&range%5Bmax%5D=20&partner.companyName=someName"`;

exports[`QuerySerializer should stringify complex filter 2`] = `
Object {
  "partner.companyName": "someName",
  "range": Object {
    "max": "20",
    "min": "10",
  },
  "test": "1",
}
`;

exports[`QuerySerializer should support url encoded values 1`] = `
Object {
  "date": Object {
    "from": "2015",
    "to": "2019",
  },
  "id": "1",
  "status": Array [
    "a",
    "b",
  ],
}
`;
